text/0000-hash-simplification.md

```diff
@@ -0,0 +1,420 @@
+
+For both this recommendation as well as the alternative below, this RFC proposes
+removing the `HashState` trait and `Hasher` structure (as well as the
+`hash_state` module) in favor of the following API:
+
+```rust
+struct HashMap<K, V, H = DefaultHasher>;
+
+impl<K: Eq + Hash, V> HashMap<K, V> {
+    fn new() -> HashMap<K, V, DefaultHasher> {
+        HashMap::with_hasher(DefaultHasher::new())
+    }
+}
+
+impl<K: Eq, V, H: Fn(&K) -> u64> HashMap<K, V, H> {
+    fn with_hasher(hasher: H) -> HashMap<K, V, H>;
+}
+
+fn global_siphash_keys() -> (u64, u64) {
+    // ...
+}
+
+impl<K: Hash> Fn(&K) -> u64 for DefaultHasher {
+    fn call(&self, arg: &K) -> u64 {
+        let (k1, k2) = global_siphash_keys();
+        let mut s = SipHasher::new_with_keys(k1, k2);
+        arg.hash(&mut s);
+        s.finish()
+    }
+}
+```
+
+The precise details will be affected based on which design in this RFC is
+chosen, but the general idea is to move from a custom trait to the standard `Fn`
+trait for calculating hashes.
+
+# Drawbacks
+
+* This design is a departure from the precedent set by many other languages. In
+  doing so, however, it is arguably easier to implement `Hash` as it's more
+  obvious how to feed in incremental state. We also do not lock ourselves into a
+  particular hashing algorithm in case we need to alternate in the future.
+
+* Implementations of `Hash` cannot be specialized and are forced to operate
+  generically over the hashing algorithm provided. This may cause a loss of
+  performance in some cases. Note that this could be remedied by moving the type
+  parameter to the trait instead of the method, but this would lead to a loss in
+  ergonomics for generic consumers of `T: Hash`.
+
+* Manual implementations of `Hash` are somewhat cumbersome still by requiring a
+  separate `Hasher` parameter which is not necessarily always desired.
+
+* The API of `Hasher` is approaching the realm of serialization/reflection and
+  it's unclear whether its API should grow over time to support more basic Rust
+  types.
+
+# Alternatives
+
+As alluded to in the "Detailed design" section the primary alternative to this
+RFC, which still improves ergonomics, is to remove the generic-ness over the
+hashing algorithm.
+
+## API
+
+The new API of `std::hash` would be:
+
+```rust
+trait Hash {
+    fn hash(&self) -> usize;
+}
+
+fn combine(a: usize, b: usize) -> usize;
+```
+
+The `Writer`, `Hasher`, and `SipHasher` structures/traits would all be removed
+from `std::hash`. This definition is more or less the Rust equivalent of the
+Java/C++ hashing infrastructure. This API is a vast simplification of what
+exists today and allows implementations of `Hash` as well as consumers of `Hash`
+to quite ergonomically work with hash values as well as hashable objects.
+
+> **Note**: The choice of `usize` instead of `u64` reflects [C++'s
+> choice][cpp-hash] here as well, but it is quite easy to use one instead of
+> the other.
+
+## Hashing algorithm
```

Comment:
Reading through this again, this is what the lingering hashing-vs-hash-coding feeling was triggered by. The Java API is a hash coding one, not a hashing one. Presumably the hashmap will hash the hash codes yielded. The C++ design, meanwhile is ambiguous to me. Its description *sounds* like hashing, but its guarantees *look* like hash coding. ([refering to this API](http://www.cplusplus.com/reference/functional/hash/))